transcendence-app/src/user/user.controller.ts

```diff
@@ -1,39 +1,63 @@
 import {
-  BadRequestException,
+  Body,
   Controller,
   Get,
   NotFoundException,
   Param,
+  ParseBoolPipe,
+  ParseIntPipe,
+  ParseUUIDPipe,
   Post,
   Query,
+  Request as GetRequest,
+  UseGuards,
 } from '@nestjs/common';
 import { UserService } from './user.service';
 import { UserEntity } from './user.entity';
 import { UserDto } from './dto/user.dto';
-import { plainToClass } from 'class-transformer';
-import { validate } from 'class-validator';
+import { AuthenticatedGuard } from '../shared/guards/authenticated.guard';
+import { Request } from 'express';
+import { ApiForbiddenResponse, ApiTags } from '@nestjs/swagger';
 
+const MAX_USER_ENTRIES_PER_PAGE = 20;
+@ApiTags('user')
+@ApiForbiddenResponse()
+@UseGuards(AuthenticatedGuard)
 @Controller('user')
 export class UserController {
   constructor(private userService: UserService) {}
 
-  @Post('new?')
-  async addUser(@Query() query: UserDto): Promise<UserEntity> {
-    const validatedUser = plainToClass(UserDto, query);
-    const errors = await validate(validatedUser, {
-      skipMissingProperties: false,
-      forbidUnknownValues: true,
-    });
-    if (errors.length > 0) {
-      throw new BadRequestException();
-    }
-    return this.userService.findOneOrCreate(query);
+  @Post()
+  async addUser(@Body() user: UserDto): Promise<UserEntity> {
+    return this.userService.findOneOrCreate(user);
   }
 
-  @Get(':id')
-  async getUserById(@Param('id') param: string): Promise<UserEntity> {
-    const result = this.userService.retrieveUserWithId(Number(param));
-    if (result === undefined) throw new NotFoundException();
+  @Get('me')
+  getCurrentUser(@GetRequest() req: Request) {
+    return req.user;
+  }
+
+  @Get('users')
+  getUsers() {
+    return this.userService.retrieveUsers();
+  }
+
+  @Get('usernames?')
+  getUsernames(
+    @Query('take', ParseIntPipe) take: number,
+    @Query('skip', ParseIntPipe) skip: number,
```

Comment:
suggestion: We could rename `take` and `skip` to `limit` and `offset`, because they are already part of [SQL](https://www.postgresql.org/docs/9.3/queries-limit.html) syntax.